libs/utils/trace_analysis.py

```diff
@@ -951,46 +846,30 @@ def plotCPUFrequencyResidency(self, cpus=None):
                           loc='lower right')
             plt.subplots_adjust(hspace=0.5)
 
-    def getClusterFrequencyResidency(self, cluster):
+    def _getClusterActiveSignal(self, cluster):
         """
-        Get a DataFrame with per cluster frequency residency, i.e. amount of
-        time spent at a given frequency in each cluster.
+        Build a square wave representing the active (i.e. non-idle) cluster
+        time, i.e.:
+            cluster_active[t] = 1 if there is a task running in the cluster at
+                                time t
+            cluster_active[t] = 0 otherwise
 
-        :param cluster: list of CPU IDs belonging to the cluster
+        :param cluster: list of CPU IDs belonging to a cluster
         :type cluster: list(int)
         """
-
-        if not self.trace.hasEvents('cpu_frequency'):
-            logging.warn('Events [cpu_frequency] not found, '\
-                         'plot DISABLED!')
-            return None
-        if not self.trace.hasEvents('cpu_idle'):
-            logging.warn('Events [cpu_idle] not found, '\
-                         'plot DISABLED!')
-            return None
-
         freq_dfr = self.trace.df('cpu_frequency')
-        cluster_freqs = freq_dfr[freq_dfr['cpu'].isin(cluster)]
+        cluster_freqs = freq_dfr[freq_dfr.cpu.isin(cluster)]
         idle_dfr = self.trace.df('cpu_idle')
 
-        ### Computer TOTAL Time ###
-        time_intervals = cluster_freqs.index[1:] - cluster_freqs.index[:-1]
-        total_time = pd.DataFrame({
-            'TOTAL Time' : time_intervals,
-            'Frequency [MHz]' : [f/1000 for f in cluster_freqs.iloc[:-1].frequency]
-        })
-        total_time = total_time.groupby(['Frequency [MHz]']).sum()
-
-        ### Compute ACTIVE Time ###
         cpu_activation = {}
         for cpu in cluster:
-            cpu_freqs = freq_dfr[freq_dfr['cpu'] == cpu]
-            cpu_idle = idle_dfr[idle_dfr['cpu_id'] == cpu]
+            cpu_freqs = freq_dfr[freq_dfr.cpu == cpu]
+            cpu_idle = idle_dfr[idle_dfr.cpu_id == cpu]
 
             # time instants where CPU goes to an idle state
-            idle_times = cpu_idle[cpu_idle['state'] != IDLE_STATE]
+            idle_times = cpu_idle[cpu_idle.state != IDLE_STATE]
             # time instants where CPU leaves an idle state
-            non_idle_times = cpu_idle[cpu_idle['state'] == IDLE_STATE]
+            non_idle_times = cpu_idle[cpu_idle.state == IDLE_STATE]
 
             if idle_times.index[0] < non_idle_times.index[0]:
                 idle_times = idle_times.iloc[1:]
```

Comment:
Is seems to me that here you are trying to force "the first IDLE_EXIT to follow the first IDLE_ENTER", is that correct?
Why do we need that?
... and here I'm lost because the code in my branch is different that what I see on that patch... :-/